gmcs/generate.py

```diff
@@ -152,6 +154,124 @@ def get_sentences_from_lkb_output(output_file,  mrs_files, verb_preds):
 def clean_tree(tree):
     return re.sub(r'\("([^()]+)"\)', r'(@\1@)', tree).replace('"', '').replace('@', "'")
 
+def generate_sentences_ace(grammar_dir, mrs_files, verb_preds, delphin_dir, session):
+    iso = os.path.basename(grammar_dir)
+
+    # command to compile the grammar
+    compile_grammar_cmd = '/usr/local/bin/ace -G %(grammar_dir)s/%(iso)s.dat -g %(grammar_dir)s/ace/config.tdl' % {'grammar_dir':grammar_dir, 'iso':iso}
+    # print(compile_grammar_cmd)
+
+    # open a file to catch the ace output
+    output = open('ace_output'+session, 'w')
+    output.write('Compiling grammar...\n> '+compile_grammar_cmd+'\n')
+    output.flush()
+
+    ace_error = open('ace_error'+session, 'w')
+    ace_error.write('Compiling grammar...\n> '+compile_grammar_cmd+'\n')
+    ace_error.flush()
+
+    # compile the grammar
+    subprocess.run(compile_grammar_cmd.split(),
+                    stdout=output, stderr=ace_error, env=os.environ)
```

Comment:
Using `str.split()` on the `compile_grammar_cmd` can fail if there are ever spaces in the `grammar_dir` directory name. I suggest doing one of the following:

1. Put quotes around your arguments when constructing the command string, then use [shlex.split()](https://docs.python.org/3/library/shlex.html#shlex.split)
2. Just make a list instead of making then splitting a string:
   ```python
   compile_cmd_args = [
       '/usr/local/bin/ace',
       '-G', f'{grammar_dir}/{iso}.dat',
       '-g', f'{grammar_dir}/ace/config.tdl',
   ]
   ```
3. Use PyDelphin's `ace.compile()`

I think (3) is the best option because it would simplify this code a lot.

Reply:
Hi Michael!

Thanks so much for your help and advice! I've made a new commit that now uses PyDelphin to call ace for both compiling and generating!

The commit is here - https://github.com/rosypen/matrix/commit/be5b86537e9cfdb72628411149d8297d9622beb9

The good news is it did clean up the code considerably and I got the parse trees, yay!

The bad news is I no longer have nicely formatted mrs because I don't think the tsdb output format does that for me. Is there a library I could use in PyDelphin for formatting the mrs?

<img width="1422" alt="Screen Shot 2022-08-13 at 1 55 20 AM" src="https://user-images.githubusercontent.com/20407310/184476675-6409ccfa-4146-4c0f-be02-ff328d12af88.png">

Reply:
> Is there a library I could use in PyDelphin for formatting the mrs?

Yes, there are options. See the list of [codecs](https://pydelphin.readthedocs.io/en/latest/api/delphin.codecs.html). If you just want SimpleMRS with indentation, that's easy:

```pycon
>>> from delphin.codecs import simplemrs
>>> m = simplemrs.decode('[ TOP: h0 INDEX: e2 [ e TENSE: pres ] RELS: < [ _rain_v_1 LBL: h1 ARG0: e2 ] > HCONS: < h0 qeq h1 > ]')
>>> print(simplemrs.encode(m, indent=True))
[ TOP: h0
  INDEX: e2 [ e TENSE: pres ]
  RELS: < [ _rain_v_1 LBL: h1 ARG0: e2 ] >
  HCONS: < h0 qeq h1 > ]
```

If you want to display DMRS or EDS instead, you'll first need to [convert the MRS](https://pydelphin.readthedocs.io/en/latest/guides/walkthrough.html#converting-semantic-representations).

And if you want a graphical web display, you can try [delphin-viz](https://github.com/delph-in/delphin-viz/).

Reply:
Thank you again!

DMRS and EDS are pending TODOs that I may get to in the future. For now this has helped me get the simple mrs formatting!

<img width="1167" alt="Screen Shot 2022-08-13 at 3 06 36 PM" src="https://user-images.githubusercontent.com/20407310/184512183-83653aa1-2762-4879-b371-72f32dd6f04c.png">